x/evm/keeper/virtual_frontier_contract.go

```diff
@@ -0,0 +1,367 @@
+			Active: activate,
+		}, &types.VFBankContractMetadata{
+			MinDenom: record.MinDenom,
+		}, &record)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DeployNewVirtualFrontierBankContract deploys a new virtual frontier bank contract into the store
+func (k Keeper) DeployNewVirtualFrontierBankContract(
+	ctx sdk.Context,
+	vfContract *types.VirtualFrontierContract,
+	bankMeta *types.VFBankContractMetadata,
+	denomMetadata *types.VirtualFrontierBankContractDenomMetadata,
+) (common.Address, error) {
+	if !denomMetadata.CanDecimalsUint8() {
+		panic(fmt.Sprintf("decimals does not fit uint8: %v", denomMetadata.Decimals))
+	}
+
+	vfContract.Type = types.VFC_TYPE_BANK
+	vfContract.Metadata = k.cdc.MustMarshal(bankMeta)
+
+	callData, err := PrepareBytecodeForVirtualFrontierBankContractDeployment(denomMetadata.Name, uint8(denomMetadata.Decimals))
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	contractAddress, err := k.DeployNewVirtualFrontierContract(ctx, vfContract, callData)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	// register mapping by denom
+	err = k.SetMappingVirtualFrontierBankContractAddressByDenom(ctx, bankMeta.MinDenom, contractAddress)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	return contractAddress, nil
+}
+
+func PrepareBytecodeForVirtualFrontierBankContractDeployment(displayName string, exponent uint8) ([]byte, error) {
+	// method is exposed to be re-use in test
+	ctorArgs, err := types.VFBankContract20.ABI.Pack(
+		"",
+		displayName,
+		displayName,
+		exponent,
+	)
+
+	if err != nil {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrPackAny, "failed to pack bytecode %s", err.Error())
+	}
+
+	contractBytecode := types.VFBankContract20.Bin
+
+	var callData []byte
+	callData = append(callData, contractBytecode...)
+	callData = append(callData, ctorArgs...)
+
+	return callData, nil
+}
+
+// DeployNewVirtualFrontierContract deploys a new virtual frontier contract into the store
+func (k Keeper) DeployNewVirtualFrontierContract(ctx sdk.Context, vfContract *types.VirtualFrontierContract, callData []byte) (contractAddress common.Address, err error) {
+	defer func() {
+		if err != nil {
+			contractAddress = common.Address{}
+		}
+	}()
+
+	if len(vfContract.Address) > 0 {
+		err = sdkerrors.ErrInvalidRequest.Wrapf("input contract address must be empty")
+		return
+	}
+
+	deployerModuleAccount := k.accountKeeper.GetModuleAccount(ctx, types.ModuleVirtualFrontierContractDeployerName)
+	if deployerModuleAccount == nil {
+		err = sdkerrors.ErrNotFound.Wrapf("module account %s does not exist", types.ModuleVirtualFrontierContractDeployerName)
+		return
+	}
+
+	nonce := deployerModuleAccount.GetSequence()
+	contractAddress = crypto.CreateAddress(types.VirtualFrontierContractDeployerAddress, nonce)
+	contractAccount := k.GetAccount(ctx, contractAddress)
+	if contractAccount != nil {
+		if len(contractAccount.CodeHash) > 0 {
+			if bytes.Equal(contractAccount.CodeHash, types.EmptyCodeHash) {
```

Comment:
not sure I understand why u check for `types.EmptyCodeHash`

Reply:
that's how to check if an account is a contract account.

I will use `if contractAccount.IsContract() {` instead, didn't realize it exists (perform the same check)